src/lib/optimizer/optimizer.cpp

```diff
@@ -204,18 +204,24 @@ void Optimizer::validate_lqp(const std::shared_ptr<AbstractLQPNode>& root_node)
         // outside of the LQP that is currently optimized. For example, if you create a bunch of LQP nodes in the test's
         // SetUp method but these are not used in the current LQP, this can cause this assertion to fail. This is only
         // enforced when rules are executed through the optimizer, not when tests call the rule directly.
-        Assert(all_nodes.contains(output), std::string{"Output `"} + output->description() + "` of node `" +
-                                               node->description() + "` not found in LQP");
+        Assert(nodes_by_lqp[lqp].contains(output), std::string{"Output `"} + output->description() + "` of node `" +
+                                                       node->description() + "` not found in LQP");
+
+        for (const auto& [other_lqp, nodes] : nodes_by_lqp) {
+          if (other_lqp == lqp) continue;
+          Assert(!nodes.contains(node), std::string{"Output `"} + output->description() + "` of node `" +
+                                                       node->description() + "` found in different LQP");
+        }
       }
 
       // Check that all LQPColumnExpressions in the node can be resolved. This does, however, not guarantee that it is
       // correctly used as one of the (transitive) inputs of `node`. Checking that would be more expensive.
```

Comment:
Should we care about that price ("be more expensive") when validate is only used in debug anyways?

Reply:
Now:

> Check that all LQPColumnExpressions in the node can be resolved. This mostly guards against expired columns leaving an optimizer rule. It does not guarantee that it the LQPColumnExpressions are correctly returned from one of the (transitive) inputs of `node`. If that is not the case, that will be caught by the LQPTranslator, at the latest. However, feel free to add that check here.

Reply:
Read: We probably could afford to do it here, but I don't feel like adding it right now ;)